lib/TPP/DecomposeConvsToMatmulOrBrgemm.cpp

```diff
@@ -0,0 +1,630 @@
+
+    // [N][H][W][C]
+    Value image = convOp.image();
+    // [R][S][C][K]
+    Value filter = convOp.filter();
+    // [N][P][Q][K]
+    Value output = convOp.getOutputs()[0];
+
+    if (!hasStaticShape(image, filter, output))
+      return failure();
+
+    FailureOr<linalg::GenericOp> maybeGeneric =
+        generalizeNamedOp(rewriter, convOp);
+    if (failed(maybeGeneric))
+      return failure();
+    linalg::GenericOp generic = *maybeGeneric;
+    generic.setLibraryCallAttr(rewriter.getStringAttr("tpp.Conv2DNhwcHwcfOp"));
+    return success();
+  }
+};
+
+// Block a Conv2DNchwFchw. The pattern returns a generic operation
+// marked as 'tpp.blocked.Conv2DNchwFchwOp' on success.
+struct BlockConv2DNchwFchw : OpRewritePattern<linalg::Conv2DNchwFchwOp> {
+  BlockConv2DNchwFchw(MLIRContext *context, ArrayRef<int64_t> blockingFactors,
+                      PatternBenefit benefit = 1)
+      : OpRewritePattern<linalg::Conv2DNchwFchwOp>(context, benefit),
+        blockingFactors(blockingFactors) {}
+
+  LogicalResult
+  blockConv2DNchwFchwPreconditions(linalg::Conv2DNchwFchwOp convOp) const {
+    // do not handle convolutions with dilation and strides.
+    if (hasStride<linalg::Conv2DNchwFchwOp>(convOp) ||
+        hasDilation<linalg::Conv2DNchwFchwOp>(convOp))
+      return failure();
+
+    // [N][C][H][W]
+    Value image = convOp.image();
+    // [K][C][R][S]
+    Value filter = convOp.filter();
+    // [N][K][P][Q]
+    Value output = convOp.getOutputs()[0];
+
+    // static shapes.
+    if (!hasStaticShape(image, filter, output))
+      return failure();
+
+    // tensor semantics.
+    if (convOp.hasBufferSemantics())
+      return failure();
+
+    // blocking factors.
+    if (blockingFactors.empty())
+      return failure();
+
+    return success();
+  }
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNchwFchwOp convOp,
+                                PatternRewriter &rewriter) const override {
+    if (failed(blockConv2DNchwFchwPreconditions(convOp)))
+      return failure();
+    FailureOr<linalg::GenericOp> maybeGeneric =
+        mlir::linalgx::packConv2DNchwFchwOp(
+            rewriter, convOp,
+            getAsOpFoldResult(rewriter.getI64ArrayAttr(blockingFactors)));
+    if (failed(maybeGeneric))
+      return failure();
+    linalg::GenericOp generic = *maybeGeneric;
+    generic.setLibraryCallAttr(
+        rewriter.getStringAttr("tpp.BlockedConv2DNchwFchwOp"));
+    return success();
+  }
+
+private:
+  SmallVector<int64_t> blockingFactors;
+};
+
+// Interchange iterator in a linalg.generic marked as
+// 'tpp.BlockedConv2DNchwFchwOp' to expose a GEMM operation.
+struct InterchangeIteratorsConv2DNchwFchw
+    : OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!tpp::utils::isMarkedWithTpp(genericOp, "tpp.BlockedConv2DNchwFchwOp"))
```

Comment:
FYI: @adam-smnk @rengolin and @KavithaTipturMadhu  this is what I dislike the most. Each pattern picks up when the other left.

Reply:
You run a canonicalization in between and you are dead.

Reply:
Worst case we can run the same pattern matcher instead of annotating, but for now we can just make sure we don't canonicalize in between.